addons/sale_purchase/models/sale_order_line.py

```diff
@@ -225,11 +225,18 @@ def _purchase_service_match_supplier(self, warning=True):
         return suppliers[0]
 
     def _purchase_service_match_purchase_order(self, partner, company=False):
-        return self.env['purchase.order'].search([
-            ('partner_id', '=', partner.id),
-            ('state', '=', 'draft'),
-            ('company_id', '=', (company and company or self.env.company).id),
-        ], order='id desc')
+        result = self.env['purchase.order.line']._read_group(
+            [
+                ('partner_id', '=', partner.id),
+                ('state', '=', 'draft'),
+                ('company_id', '=', (company and company or self.env.company).id),
+                ('sale_order_id', '=', self.order_id.id),
+            ],
+            ['order_id'],
+            order='order_id',
+            limit=1,
+        )
+        return result[0][0] if result else []
```

Comment:
This returns a recordset if something is found, and a list otherwise. This may lead to unexpected behavior. :confused: 
Is there a reason to use a `_read_group` to search for a single record?

Reply:
> This returns a recordset if something is found, and a list otherwise. This may lead to unexpected behavior. 😕 Is there a reason to use a `_read_group` to search for a single record?

I replaced the empty list with the purchase order object. Xbo has explained the reason in this comment. [link](https://github.com/odoo/odoo/pull/176451#discussion_r1873518480)

Reply:
I really fail to see how adding a `group by` could speed up the process. :confused: 
Imho you should search on the first `purchase.order.line` that matches your criteria and grab its order_id.

Reply:
Ah true indeed